In [1]:
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import log_loss
from sklearn import preprocessing
from sklearn.preprocessing import scale, MinMaxScaler

In [2]:
df = pd.read_csv(open("data_v1.csv", "r", encoding='cp949='))

In [3]:
print(df.shape)
print(df.columns)
df.head()

(1905, 34)
Index(['Unnamed: 0', '국가코드', 'DB종류', '특허/실용 구분', '문헌종류 코드', '발명의 명칭', '요약',
       '대표청구항', '청구항 수', '출원번호', '출원일', '공개번호/공표/재공표 ', '공개일', '등록번호', '등록일',
       '출원인', '출원인 국적', '출원인 수', '출원인 대표명화 코드', '출원인 대표명화 영문명', '발명자/고안자',
       '우선권 번호', '우선권 국가', '우선권 주장일', '국제 공개번호', '국제 공개일', 'Original CPC Main',
       'Original CPC All', 'Original IPC Main', 'Original IPC All',
       'Original US Class Main[US]', 'Original US Class All[US]',
       'WIPS ON key', '유효특허 여부'],
      dtype='object')


Unnamed: 0 국가코드 DB종류 특허/실용 구분 문헌종류 코드  \
0           0   US   US        P      B2   
1           1   US   US        P      A1   
2           2   US   US        P      A1   
3           3   US   US        P      A1   
4           4   US   US        P      B2   

                                              발명의 명칭  \
0                      Turboengine water wash system   
1  TREATMENT OF TAILINGS WITH DEIONIZED SILICATE ...   
2  Method And Apparatus For Treating Tailings Usi...   
3   METHOD FOR STABILIZING WASTE AND HAZARDOUS WASTE   
4  Water treatment process and plant using ballas...   

                                                  요약  \
0  A system for cleaning gas turbine engines is d...   
1  A process for treating a tailings stream compr...   
2  There is a method described of treating tailin...   
3  The invention provides a process for stabilizi...   
4  A method for treating wastewater using a balla...   

                                               대표청구항  청구항 수       출원번호  \
0  1. An apparatus for cleaning wing mounted gas ...     64  11/644784   
1  1. A process for treating a tailings stream co...     20  13/848244   
2  1. A method of treating tailings having a comb...     18  13/765924   
3  1. A method for stabilizing hazardous waste, c...     20  13/767408   
4  1. A method of treating wastewater including: ...     18  12/521154   

    ...          국제 공개번호      국제 공개일 Original CPC Main  \
0   ...              NaN         NaN               NaN   
1   ...              NaN         NaN    B09B-0003/0025   
2   ...              NaN         NaN      C02F-0001/48   
3   ...              NaN         NaN      A62D-0003/33   
4   ...    WO2008-083923  2008-07-17               NaN   

               Original CPC All Original IPC Main  \
0                           NaN             20802   
1                B09B-0003/0025             20902   
2                  C02F-0001/48             30206   
3  A62D-0003/33 | G21F-0009/302             16204   
4                           NaN             30206   

                Original IPC All Original US Class Main[US]  \
0                          20802                  134/166.R   
1                          20902                    106/627   
2                          30206                    204/554   
3                    16204 72106                    588/002   
4  30206 20104 20104 20104 20104                    210/709   

                           Original US Class All[US]   WIPS ON key 유효특허 여부  
0                                134/166.R | 134/138  4.914000e+12     0.0  
1  106/627 | 106/600 | 106/631 | 106/632 | 106/63...  5.414000e+12     0.0  
2                                            204/554  5.414000e+12     0.0  
3                                  588/002 | 588/315  5.414000e+12     0.0  
4  210/709 | 210/711 | 210/712 | 210/713 | 210/71...  4.913050e+12     0.0  

[5 rows x 34 columns]

In [4]:
features = ['공개번호/공표/재공표 ', '등록번호', 'Original IPC Main']
x = df[features]
y = df['유효특허 여부']
x = x.fillna(value=0)
x.head()

공개번호/공표/재공표      등록번호  Original IPC Main
0  2.008015e+10  8628627              20802
1  2.014001e+10        0              20902
2  2.014001e+10        0              30206
3  2.014001e+10        0              16204
4  2.010010e+10  8617400              30206

In [5]:
x.astype(int).astype('str').head()

공개번호/공표/재공표      등록번호 Original IPC Main
0  20080149141  8628627             20802
1  20140007795        0             20902
2  20140008229        0             30206
3  20140005461        0             16204
4  20100096335  8617400             30206

In [6]:
le_1 = preprocessing.LabelEncoder()
le_1.fit(x['공개번호/공표/재공표 '])
x['공개번호/공표/재공표 '] = le_1.transform(x['공개번호/공표/재공표 '])
x['공개번호/공표/재공표 '].head()

0     630
1    1096
2    1097
3    1095
4     773
Name: 공개번호/공표/재공표 , dtype: int64

In [7]:
# le_2 = preprocessing.LabelEncoder()
# le_2.fit(x['등록번호'])
# x['등록번호'] = le_2.transform(x['등록번호'])
# x['등록번호'].head()

In [8]:
# le_3 = preprocessing.LabelEncoder()
# le_3.fit(x['Original IPC Main'])
# x['Original IPC Main'] = le_3.transform(x['Original IPC Main'])
# x['Original IPC Main'].head()

In [9]:
X_train, X_val, y_train, y_val = train_test_split(x, y, test_size=0.33)
clf = RandomForestClassifier(n_estimators=10)
clf.fit(X_train, y_train)
y_val_pred = clf.predict_proba(X_val)
log_loss(y_val, y_val_pred)

0.34667846624285131

공개번호 -> 0.2<br>
공개번호, IPC -> 0.2~0.4<br>
공개번호, 등록번호 -> 0.35 <br>
공개번호, 등록번호, IPC -> 0.3~0.4